akka-stream/src/main/scala/akka/stream/javadsl/SinkRef.scala

```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.javadsl
+
+import java.util.concurrent.CompletionStage
+
+import akka.stream.scaladsl.{ SinkRef, SourceRef }
+import akka.stream.stage.GraphStageWithMaterializedValue
+import akka.util.OptionVal
+import akka.stream.{ SinkShape, scaladsl }
+
+import scala.concurrent.Future
+
+object SinkRef {
+
+  import scala.compat.java8.FutureConverters._
+
+  // TODO this is hell to do since no flatMapMaterialized
+  //  def source[T](): Source[T, CompletionStage[>>> the java one SinkRef[T]]] =
+  //    scaladsl.SinkRef.source[T]()
+  //      .mapMaterializedValue(_.toJava) // still have the scala one
+  //      .asJava
+
+  def source[T](): Source[T, CompletionStage[SinkRef[T]]] =
+    ???
+  //    scaladsl.SinkRef.source[T]()
+  //      .mapMaterializedValue(_.toJava)
+  //      .asJava
+
+}
+
+abstract class SinkRef[In] extends GraphStageWithMaterializedValue[SinkShape[In], Future[SourceRef[In]]] with Serializable {
```

Comment:
As mentioned, pulling off the `javadsl` / `scaladsl` is very hairy so I abbandoned it eventually.

The problem is the materialized value really:

- The both javadsl and scaladsl have to inherit eachother, in order to allow people to `match` them easily, since those will be passed around actors etc
  - we must not bind "the other side was using scala, so you have to as well!
- If we do `javadsl` being `abstract class` with Mat Value: `CompletionStage[javadsl.SourceRef[In]]`... then `scaladsl` can't really extend it... since it needs `Future[scaladsl.SourceRef`

Summing up, I think it's safer to go with 1 type -- put them all in `akka.stream.SourceRef` and `akka.stream.SinkRef`. The only place where it makes a difference really is when people receive a SourceRef, and need to call `getSource` rather than `.source` on it. Notice that for the sink it's very nice, since regardless of language, doing `runWith(theSinkSource)` always for "remote" (consumers), since they *do not want the materialized value anyway!* -- the Future there is forced to blow up (the "loop materializations avoidance")

// @jrudolph the 2nd crazy topic;

Reply:
Put a proposal into https://github.com/ktoso/akka/pull/33